tests/test_inheritance.cpp

```diff
@@ -97,4 +97,27 @@ test_initializer inheritance([](py::module &m) {
             py::isinstance<Unregistered>(l[6])
         );
     });
+
+    m.def("test_matching_holder_type", []() {
+        struct Base { };
+        struct Derived : Base { };
+
+        try {
+            {
+                py::module m;
+                py::class_<Base, std::shared_ptr<Base>>(m, "Base");
+                py::class_<Derived, Base>(m, "Derived")
+                    .def(py::init<>());
+            }
+            {
+                py::module m;
+                py::class_<Base>(m, "Base");
+                py::class_<Derived, std::shared_ptr<Derived>, Base>(m, "Derived")
+                    .def(py::init<>());
+            }
+        } catch (std::runtime_error &) {
+            return true;
```

Comment:
Sorry, I need to revise my original suggestion a bit. It might be nicer to catch the exception in Python in order to also check that the correct exception was thrown.

```c++
m.def("test_matching_holder_type", []() {
    struct Base { };
    struct Derived : Base { };

    py::class_<Base>(m, "Base");
    py::class_<Derived, std::shared_ptr<Derived>, Base>(m, "Derived");
});
```
```python
with pytest.raises(RuntimeError) as excinfo:
    test_matching_holder_type()
assert str(excinfo.value) == "generic_type: ..."
```